webui/react/src/e2e/models/hew/DataGrid.ts

```diff
@@ -0,0 +1,330 @@
+> = NamedComponentArgs & { parentTable: DataGrid<RowType, HeadRowType> };
+export type HeadRowArgs = NamedComponentArgs & { clickableParentLocator: Locator };
+export type TableArgs<
+  RowType extends Row<RowType, HeadRowType>,
+  HeadRowType extends HeadRow,
+> = NamedComponentArgs & {
+  rowType: RowClass<RowType, HeadRowType>;
+  headRowType: HeadRowClass<HeadRowType>;
+};
+
+/**
+ * Returns a representation of the DataGrid component.
+ * This constructor represents the contents in hew/src/kit/DataGrid.tsx.
+ * @param {object} obj
+ * @param {CanBeParent} obj.parent - The parent used to locate this DataGrid
+ * @param {string} [obj.selector] - Used instead of `defaultSelector`
+ * @param {RowType} [obj.rowType] - Value for the RowType used to instanciate rows
+ * @param {HeadRowType} [obj.headRowType] - Value of the HeadRowType used to instanciate the head row
+ */
+export class DataGrid<
+  RowType extends Row<RowType, HeadRowType>,
+  HeadRowType extends HeadRow,
+> extends NamedComponent {
+  readonly defaultSelector: string = '[class^="DataGrid_base"]';
+  constructor(args: TableArgs<RowType, HeadRowType>) {
+    super(args);
+    this.#rowType = args.rowType;
+    this.rows = new args.rowType({
+      parent: this.#body,
+      parentTable: this,
+    });
+    this.headRow = new args.headRowType({
+      clickableParentLocator: this.pwLocator,
+      parent: this.#head,
+    });
+  }
+  readonly canvasTable: BaseComponent = new BaseComponent({
+    parent: this,
+    selector: 'canvas[data-testid="data-grid-canvas"] table',
+  });
+  readonly #otherCanvas: BaseComponent = new BaseComponent({
+    parent: this,
+    selector: 'canvas:not([data-testid])',
+  });
+  #columnheight: number | undefined;
+  readonly #rowType: RowClass<RowType, HeadRowType>;
+  readonly rows: RowType;
+  readonly headRow: HeadRowType;
+  readonly #body: BaseComponent = new BaseComponent({
+    parent: this.canvasTable,
+    selector: 'tbody',
+  });
+  readonly #head: BaseComponent = new BaseComponent({
+    parent: this.canvasTable,
+    selector: 'thead',
+  });
+
+  get columnHeight(): number {
+    if (this.#columnheight === undefined || Number.isNaN(this.#columnheight)) {
+      throw new Error('Please use setColumnHeight to set the column height');
+    }
+    return this.#columnheight;
+  }
+
+  /**
+   * Returns the height of the header column. Trying to click on a row will fail
+   * if this method hasn't yet been run.
+   */
+  async setColumnHeight(): Promise<number> {
+    const style = await this.#otherCanvas.pwLocator.getAttribute('style');
+    if (style === null) {
+      throw new Error("Couldn't find style attribute.");
+    }
+    const matches = style.match(/height: (\d+)px;/);
+    if (matches === null) {
+      throw new Error("Couldn't find height in style attribute.");
+    }
+    this.#columnheight = +matches[1];
+    return this.columnHeight;
+  }
+
+  /**
+   * Returns a row from an index. Start counting at 1.
```

Comment:
why are we doing 1-based indexing?

Reply:
Datagrid starts counting at 1 for it's head row. And row 2 is the first row of the body.

Reply:
Actually, you're right. I can just use n+2